### PR TITLE
feat: async ccc checker

### PR DESCRIPTION
**Pull request by omerfirmak**
## 1. Purpose or design rationale of this PR

this has been in testing for some time already, I want to merge it to develop and start deploying to internal follower nodes.

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [x] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [x] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [x] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

File: cmd/utils/flags.go [hunk 1]
```diff
@@ -847,6 +848,12 @@ var (
 		Usage: "Enable circuit capacity check during block validation",
 	}
 
+	CircuitCapacityCheckWorkersFlag = cli.UintFlag{
```
Comment (Thegaram) [inline cmd/utils/flags.go:851]: Fyi you have to add this in another file (forgot which) if you want it to be printed with `geth --help`.

File: rollup/ccc/async_checker.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+}
+
+// Check spawns an async CCC verification task.
+func (c *AsyncChecker) Check(block *types.Block) error {
+	if block.NumberU64() > c.currentHead.Number.Uint64()+1 {
+		log.Error("non continuous chain observed in AsyncChecker", "prev", c.currentHead, "got", block.Header())
+	} else if block.ParentHash() != c.currentHead.Hash() {
+		// seems like there is a fork happening, a block from the canonical chain must have failed CCC check
+		// assume the incoming block is the new tip in the fork
+		c.forkCtx, c.forkCtxCancelFunc = context.WithCancel(context.Background())
```
Comment (Thegaram) [inline rollup/ccc/async_checker.go:101]: Does this line implicitly cancel the previous fork context?

File: core/block_validator.go [hunk 3]
```diff
@@ -114,25 +96,11 @@ func (v *BlockValidator) ValidateBody(block *types.Block) error {
...
-		// (e.g., in miner/worker.go or in insertChain),
-		// we simply skip its calculation and validation
-		if rawdb.ReadBlockRowConsumption(v.bc.db, block.Hash()) != nil {
-			return nil
-		}
-		rowConsumption, err := v.validateCircuitRowConsumption(block)
-		if err != nil {
+
+	if v.asyncValidator != nil {
+		if err := v.asyncValidator(block); err != nil {
```
Comment (Thegaram) [inline core/block_validator.go:103]: Might be interesting to add a time metric here (or inside `Check`), to see how often and for how long this call is blocking.

**Review by Thegaram — DISMISSED** [review]

File: rollup/ccc/async_checker.go (hunk 2 above)
Comment (omerfirmak) [reply rollup/ccc/async_checker.go:101]: Nope, it doesnt. But if we are observing a fork here, it is most likely that previous fork context was canceled already.

File: cmd/utils/flags.go (hunk 1 above)
Comment (omerfirmak) [reply cmd/utils/flags.go:851]: thanks, done in https://github.com/scroll-tech/go-ethereum/pull/996/commits/3e692e4cd03d6972501fe42b001c25d8107a45f7

File: core/block_validator.go (hunk 3 above)
Comment (omerfirmak) [reply core/block_validator.go:103]: I had something similar in worker but not here, thanks! added in https://github.com/scroll-tech/go-ethereum/pull/996/commits/0a8132aa7aee795f64f5bf0de54c25849ac752f1

**Review by Thegaram — DISMISSED** [review]

**Review by 0xmountaintop — DISMISSED** [review]

**Review by 0xmountaintop — APPROVED** [review]

**Review by colinlyguo — APPROVED** [review]

**Pull request merged**